examples/vault-ec2-auth/user-data-auth-client.sh

```diff
@@ -0,0 +1,99 @@
+readonly pkcs7=$(curl -s http://169.254.169.254/latest/dynamic/instance-identity/pkcs7 | tr -d '\n')
+readonly data=$(cat <<EOF
+{
+  "role": "${example_role_name}",
+  "pkcs7": "$pkcs7"
+}
+EOF
+)
+
+# run-consul is running on the background, so in case it fails we retry
+for i in $(seq 1 10); do login_output=$(curl --request POST --data "$data" "https://vault.service.consul:8200/v1/auth/aws/login") && s=0 && break || s=$? && sleep 20; done; (exit $s)
+
+# It is important to note that the default behavior is TOFU(trust on first use)
+# So if the pkcs7 certificate gets compromised, attempts to login again will be
+# denied unless the client "nonce" returned at the first login is also provided
+# Read more at https://www.vaultproject.io/docs/auth/aws.html#client-nonce
+#
+# nonce=$(echo $login_output | jq -r .auth.metadata.nonce)
+# readonly data=$(cat <<EOF
+# {
+#   "role": "${example_role_name}",
+#   "pkcs7": "$pkcs7",
+#   "nonce": "$nonce"
+# }
+# EOF
+# )
+# curl --request POST --data "$data" "https://vault.service.consul:8200/v1/auth/aws/login"
+#
+# ==============================================================================
+# The output after login will be similar to this:
+# {
+#   "request_id": "eed334ef-30bc-44a4-2a7f-93ecd7ce23cd",
+#   "lease_id": "",
+#   "renewable": false,
+#   "lease_duration": 0,
+#   "data": null,
+#   "wrap_info": null,
+#   "warnings": [
+#     "TTL of \"768h0m0s\" exceeded the effective max_ttl of \"500h0m0s\"; TTL value is capped accordingly"
+#   ],
+#   "auth": {
+#     "client_token": "0ac5b97d-9637-9c03-ce37-77565ed66b8a",
+#     "accessor": "f56d56cf-b3a9-d77b-439e-5ea42563a62b",
+#     "policies": [
+#       "default",
+#       "example-policy"
+#     ],
+#     "token_policies": [
+#       "default",
+#       "example-policy"
+#     ],
+#     "metadata": {
+#       "account_id": "738755648600",
+#       "ami_id": "ami-0a50e8de57a8606a7",
+#       "instance_id": "i-0e1c0ef82afa24a7c",
+#       "nonce": "d60cf363-eb83-3142-74c3-647445365e32",
+#       "region": "eu-west-1",
+#       "role": "dev-role",
+#       "role_tag_max_ttl": "0s"
+#     },
+#     "lease_duration": 1800000,
+#     "renewable": true,
+#     "entity_id": "5051f586-eef5-064e-eca6-768b1de7d19f"
+#   }
+# }
+
+# We can then use the client token from this output
+readonly token=$(echo $login_output | jq -r .auth.client_token)
+
+# And use the token to perform operations on vault such as reading a secret
+readonly response=$(curl \
+  -H "X-Vault-Token: $token" \
+  -X GET \
+  https://vault.service.consul:8200/v1/secret/example_gruntwork)
+
+# If vault cli is installed we can also perform this operation by setting the necessary environment variables
+# export VAULT_TOKEN=$token
+# export VAULT_ADDR=https://vault.service.consul:8200
+# /opt/vault/bin/vault read secret/example_gruntwork
+
+# Serves the answer in a web server
+echo $response | jq -r .data.the_answer > index.html
+python -m SimpleHTTPServer 8080 &
```

Comment:
A couple thoughts:

1. I wouldn't recommend Python's `SimpleHTTPServer`, as it crashes/hangs with concurrent requests. With Ubuntu, we usually use busybox's built-in httpd instead: https://github.com/brikis98/terraform-up-and-running-code/blob/master/code/bash/03-terraform-state/user-data.sh#L9
1. I remember having to add `nohup` to prevent the server from exiting when the User Data script exits... But I'm not 100% sure that's necessary.

Reply:
1. I don't think this would be an issue here because, it is a simple example where the server is only there to expose one value and there wouldn't be concurrent requests coming in from the tests. Also this would work on Amazon Linux too. 

2. I considered adding nohup, but in the end it wasn't necessary

Reply:
> I don't think this would be an issue here because, it is a simple example where the server is only there to expose one value and there wouldn't be concurrent requests coming in from the tests. Also this would work on Amazon Linux too.

All it takes is for one extra health check (e.g., from an ELB) or a single manual request while a test is running, and it'll crash... Not worth the potential for extra confusion IMO.

Reply:
Fair enough. But what about Amazon Linux then?

Reply:
There are lots of options: https://gist.github.com/willurd/5720255. I can't remember which we typically use for Amazon Linux. Perhaps Ruby is installed by default?

```
ruby -run -ehttpd . -p8000
```

Reply:
ruby isn't preinstalled in the ubuntu machine